services/src/main/scala/cromwell/pipeline/service/ProjectConfigurationService.scala

```diff
@@ -1,17 +1,49 @@
 package cromwell.pipeline.service
 
 import cromwell.pipeline.datastorage.dao.repository.DocumentRepository
-import cromwell.pipeline.datastorage.dto.{ ProjectConfiguration, ProjectId }
+import cromwell.pipeline.datastorage.dto.{
+  FileParameter,
+  ProjectConfiguration,
+  ProjectFileConfiguration,
+  ProjectId,
+  TypedValue
+}
 
+import java.nio.file.Path
 import scala.concurrent.{ ExecutionContext, Future }
 
 class ProjectConfigurationService(repository: DocumentRepository)(implicit ec: ExecutionContext) {
   import ProjectConfiguration._
 
-  def addConfiguration(projectConfiguration: ProjectConfiguration): Future[String] =
-    repository
-      .updateOne(toDocument(projectConfiguration), "projectId", projectConfiguration.projectId.value)
-      .map(_.toString)
+  def addConfiguration(projectConfiguration: ProjectConfiguration): Future[String] = {
+    def updateConfiguration(configuration: ProjectConfiguration) =
+      repository.updateOne(toDocument(configuration), "projectId", projectConfiguration.projectId.value).map(_.toString)
+
+    def toMap(config: ProjectFileConfiguration): Map[Path, Map[String, TypedValue]] =
+      Map(
+        config.path -> config.inputs.flatMap(fileParameter => Map(fileParameter.name -> fileParameter.typedValue)).toMap
+      )
+
+    def toProjectFileConfiguration(path: Path, inputs: Map[String, TypedValue]): ProjectFileConfiguration =
+      ProjectFileConfiguration(path, inputs.flatMap {
+        case (name, typedValue) => List(FileParameter(name, typedValue))
+      }.toList)
+
+    getById(projectConfiguration.projectId).flatMap {
+      case Some(config) if projectConfiguration.projectFileConfigurations != config.projectFileConfigurations =>
+        val oldFileConfigsMap = config.projectFileConfigurations.flatMap(toMap).toMap
+        val newFileConfigsMap = projectConfiguration.projectFileConfigurations.flatMap(toMap).toMap
+
+        val updatedFileConfigsMap = oldFileConfigsMap ++ newFileConfigsMap
+        val updatedFileConfigs = updatedFileConfigsMap.flatMap {
+          case (path, inputs) => List(toProjectFileConfiguration(path, inputs))
+        }.toList
+
+        val newConfig = projectConfiguration.copy(projectFileConfigurations = updatedFileConfigs)
+        updateConfiguration(newConfig)
+      case _ => updateConfiguration(projectConfiguration)
```

Comment:
Sorry for not bringing it up earlier but are we sure we want to update a configuration if we already know that old and new configurations are identical? I mean this default case can be achieved either if there is no configuration or if there is a configuration but it's the same as we want to put. Maybe we should not do anything in this case? Alternatively you case remove guardian condition in the first `case`